File: data-pipeline/ep-core/src/main/java/org/ekstep/ep/samza/util/DeDupEngine.java
package org.ekstep.ep.samza.util;


import redis.clients.jedis.Jedis;
import redis.clients.jedis.exceptions.JedisException;

public class DeDupEngine {

    private RedisConnect redisConnect;
    private Jedis redisConnection;
    private int store;
    private int expirySeconds;

    public DeDupEngine(RedisConnect redisConnect, int store, int expirySeconds) {
      
        this.redisConnect = redisConnect;
        this.redisConnection = redisConnect.getConnection();
        this.store = store;
        this.redisConnection.select(store);
        this.expirySeconds = expirySeconds;
    }

    public boolean isUniqueEvent(String checksum) throws JedisException {

        boolean unique = false;
        try {
            unique = !redisConnection.exists(checksum);
        } catch (JedisException ex) {
            this.redisConnection.close();
            this.redisConnection = redisConnect.getConnection(store, 10000);
            unique = !redisConnection.exists(checksum);
        }
        return unique;
    }

    public void storeChecksum(String checksum) throws JedisException {

        try {
            redisConnection.setex(checksum, expirySeconds, "");
        } catch (JedisException ex) {
            this.redisConnection.close();
            this.redisConnection = redisConnect.getConnection(10000);
            this.redisConnection.select(store);
            redisConnection.setex(checksum, expirySeconds, "");
        }
    }

    public Jedis getRedisConnection() {
        return redisConnection;
    }

}

File: data-pipeline-flink/dp-core/src/main/scala/org/sunbird/dp/core/cache/DedupEngine.scala
package org.sunbird.dp.core.cache

import redis.clients.jedis.Jedis
import redis.clients.jedis.exceptions.JedisException


class DedupEngine(redisConnect: RedisConnect, store: Int, expirySeconds: Int) extends Serializable {

  private val serialVersionUID = 6089562751616425354L
  private[this] var redisConnection: Jedis = redisConnect.getConnection
  redisConnection.select(store)

  @throws[JedisException]
  def isUniqueEvent(checksum: String): Boolean = {
    var unique = false
    try {
      unique = !redisConnection.exists(checksum)
    } catch {
      case ex: JedisException =>
        ex.printStackTrace()
        this.redisConnection.close()
        this.redisConnection = redisConnect.getConnection(this.store, backoffTimeInMillis = 10000)
        unique = !this.redisConnection.exists(checksum)
    }
    unique
  }

  @throws[JedisException]
  def storeChecksum(checksum: String): Unit = {
    try
      redisConnection.setex(checksum, expirySeconds, "")
    catch {
      case ex: JedisException =>
        ex.printStackTrace()
        this.redisConnection.close()
        this.redisConnection = redisConnect.getConnection(this.store, backoffTimeInMillis = 10000)
        this.redisConnection.select(this.store)
        this.redisConnection.setex(checksum, expirySeconds, "")
    }
  }

  def getRedisConnection: Jedis = redisConnection

  def closeConnectionPool(): Unit = {
    redisConnection.close()
  }
}
